### PR TITLE
Restore the ability to add custom schema and configuration sources when constructing an executable

### PR DESCRIPTION
**Pull request by BrynCooke**
You and now specify custom schema and config sources when constructing an executable.
```rust
Executable::builder()
  .shutdown(ShutdownSource::None)
  .schema(SchemaSource::Stream(schemas))
  .config(ConfigurationSource::Stream(configs))
  .start()
  .await
```

### CODE REVIEW

File: NEXT_CHANGELOG.md [hunk 1]
```diff
@@ -65,6 +65,18 @@ telemetry:
 By [@BrynCooke](https://github.com/BrynCooke) & [@bnjjj](https://github.com/bnjjj) & [@o0Ignition0o](https://github.com/o0Ignition0o) in https://github.com/apollographql/router/pull/1514
 
 
+### Restore the ability to specify custom schema and configuration sources ([#1733](https://github.com/apollographql/router/issues/1733))
+You and now specify custom schema and config sources when constructing an executable.
```
Comment (garypen) [inline NEXT_CHANGELOG.md:69]:
```suggestion
You may now specify custom schema and config sources when constructing an executable.
```

**Review by bnjjj — APPROVED** [review]

**Review by garypen — COMMENTED** [review]
Would it be neater to re-structure this so that we add an alternative start method (`start_with_schema_and_config`) and just basically avoid the runtime check for clashing of mechanisms and `Option`s ?

**Comment by BrynCooke** [comment]
Maybe, but in this case we still want the router to react in the same way to all the regular Opts, including env variables.

**Comment by garypen** [comment]
> Maybe, but in this case we still want the router to react in the same way to all the regular Opts, including env variables.

I'm not sure that's true though. We explicitly don't want it to respect `--config` or `-s` if we provide them via the API. By having a method that doesn't respect those, we remove that complex matching/runtime checking and we can document that in the new function.

I don't think this is a big deal, but the current solution does seem a bit cluttered.

**Comment by SimonSapin** [comment]
> Maybe, but in this case we still want the router to react in the same way to all the regular Opts, including env variables.

My reaction to the PR title and description was: does `RouterHttpServer` not satisfy the use case of providing things programmatically rather than CLI options?

But it sounds like you want *some* CLI options to apply?

**Comment by BrynCooke** [comment]
Yes, we want pretty much all the standard behaviour around CLI.
Going the `RouterHttpServer` route will mean creating a large amount of new CLI code that will have to stay in sync.

**Comment by BrynCooke** [comment]
> I'm not sure that's true though. We explicitly don't want it to respect `--config` or `-s` if we provide them via the API. By having a method that doesn't respect those, we remove that complex matching/runtime checking and we can document that in the new function.
> 
> I don't think this is a big deal, but the current solution does seem a bit cluttered.

It's not ideal, but I do want Opts to be parsed even if we add a new method. So the runtime check would have to be there regardless.

**Comment by BrynCooke** [comment]
@garypen I had a look at `Clap` and although there is a `flatten` option I started implementing with it and things were just getting more complicated.

In my case I want to override schema but not config, so having the extra method still isn't good enough as we would need permutations for all the possibilities.

I think the runtime check is the least worst option.

**Review by garypen — APPROVED** [review]

**Pull request merged**